#! /bin/sh /usr/share/dpatch/dpatch-run
## unpartitioned-disks.dpatch by Colin Watson <cjwatson@ubuntu.com>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Don't try to call BLKPG ioctls on disks that can't be partitioned.

@DPATCH@
diff -urNad parted-1.7.1~/libparted/arch/linux.c parted-1.7.1/libparted/arch/linux.c
--- parted-1.7.1~/libparted/arch/linux.c	2007-09-07 11:35:42.000000000 +0100
+++ parted-1.7.1/libparted/arch/linux.c	2007-09-07 11:59:52.000000000 +0100
@@ -1771,6 +1771,18 @@
 }
 
 static int
+_has_partitions (const PedDisk* disk)
+{
+        PED_ASSERT(disk != NULL, return 0);
+
+        /* Some devices can't be partitioned. */
+        if (!strcmp (disk->type->name, "loop"))
+                return 0;
+
+        return 1;
+}
+
+static int
 _blkpg_part_command (PedDevice* dev, struct blkpg_partition* part, int op)
 {
         LinuxSpecific*          arch_specific = LINUX_SPECIFIC (dev);
@@ -1794,6 +1806,9 @@
         PED_ASSERT(disk != NULL, return 0);
         PED_ASSERT(disk->dev->sector_size % 512 == 0, return 0);
 
+        if (!_has_partitions (disk))
+                return 0;
+
         if (ped_disk_type_check_feature (disk->type,
                                          PED_DISK_TYPE_PARTITION_NAME))
                 vol_name = ped_partition_get_name (part);
@@ -1838,6 +1853,9 @@
 {
         struct blkpg_partition  linux_part;
 
+        if (!_has_partitions (disk))
+                return 0;
+
         memset (&linux_part, 0, sizeof (linux_part));
         linux_part.pno = n;
         return _blkpg_part_command (disk->dev, &linux_part,
